src/kernel/include/shared/util.h

```diff
@@ -0,0 +1,67 @@
+#ifndef UTIL_H
```

Comment:
Qi should already have definitions for many of these.  Talk to him about where he put them, and what they're called.  `util.h` is too generic a name.  Currently the file should be called `atomic_sync.h`, I'd imagine.